### PR TITLE
Fix i18n _get_default_locale_path

### LINKED ISSUE #105: Some tests fails with pytest-randomly
### What did you do?

I ran tests with `pytest-randomly` plugin installed.

### What did you expect to happen?

All tests pass.

### What actually happened?

Few tests failed.

### What versions are you using?

* OS: OpenIndiana
* Python: 3.9.16
* Humanize: 4.6.0

Example of failed tests:

```
=========================== short test summary info ============================
FAILED tests/test_i18n.py::test_ordinal_genders[es_ES-1-male-1\xba] - Excepti...
FAILED tests/test_i18n.py::test_intword_plurals[es_ES-1000000000-1.0 bill\xf3n]
FAILED tests/test_i18n.py::test_langauge_codes[ar-5\u062e\u0627\u0645\u0633]
FAILED tests/test_i18n.py::test_intword_plurals[es_ES-1000000000000-1.0 trill\xf3n]
FAILED tests/test_i18n.py::test_langauge_codes[fr-5e] - Exception: Humanize c...
FAILED tests/test_i18n.py::test_langauge_codes[ar_SA-5\u062e\u0627\u0645\u0633]
FAILED tests/test_i18n.py::test_ordinal_genders[fr_FR-1-male-1er] - Exception...
FAILED tests/test_i18n.py::test_ordinal_genders[es_ES-5-female-5\xaa] - Excep...
FAILED tests/test_i18n.py::test_intword_plurals[es_ES-6700000000000-6.7 trillones]
FAILED tests/test_i18n.py::test_intword_plurals[es_ES-1000000-1.0 mill\xf3n]
FAILED tests/test_i18n.py::test_langauge_codes[pt_BR-5\xba] - Exception: Huma...
FAILED tests/test_i18n.py::test_ordinal_genders[it_IT-8-female-8\xaa] - Excep...
FAILED tests/test_i18n.py::test_naturaldelta - Exception: Humanize cannot det...
FAILED tests/test_i18n.py::test_intcomma - Exception: Humanize cannot determi...
FAILED tests/test_i18n.py::test_langauge_codes[pt-5\xba] - Exception: Humaniz...
```

Please note the list and number of failed tests differs with every test run.

### PR DESCRIPTION
**Pull request by dangillet**
Fixes #105

Changes proposed in this pull request:

* Use `importlib.resources` available since Python 3.9 to find the location of the locale folder containing the translations.

### CODE REVIEW

**Comment by dangillet** [comment]
I tried to use `__package__` initially, but the docs suggest to use `__spec__.parent` instead. <https://docs.python.org/3/reference/import.html#package__>

The tests cover when `__spec__` is `None` or even if it's missing. As `__spec__.parent` [is read-only ](https://docs.python.org/3/library/importlib.html#importlib.machinery.ModuleSpec.parent), I cannot make a test which sets a different value for it.

File: src/humanize/i18n.py [hunk 1]
```diff
@@ -32,14 +34,14 @@
...
-        if __file__ is None:
-            return None
-        return os.path.join(os.path.dirname(__file__), "locale")
-    except NameError:
+def _get_default_locale_path() -> pathlib.Path | None:
+    package = __spec__ and __spec__.parent
+    if not package:
         return None
 
+    with importlib.resources.as_file(importlib.resources.files(package)) as pkg:
```
Comment (hugovk) [inline src/humanize/i18n.py:42]:
```suggestion
    import importlib.resources

    with importlib.resources.as_file(importlib.resources.files(package)) as pkg:
```

File: src/humanize/i18n.py [hunk 2]
```diff
@@ -3,7 +3,9 @@
 from __future__ import annotations
 
 import gettext as gettext_module
-import os.path
+import importlib.resources
+import os
+import pathlib
 from threading import local
```
Comment (hugovk) [inline src/humanize/i18n.py:9]:
```suggestion
from threading import local
from typing import TYPE_CHECKING

if TYPE_CHECKING:
    import os
    import pathlib
```

**Review by hugovk — COMMENTED** [review]
Thanks for this!

Let's do some tuning to speed up with lazy imports, which can be important for CLIs where you want a quick response time.

Using https://github.com/nschloe/tuna we can visualise the import times:

```sh
python -m pip install tuna
python -c "import humanize" && python -X importtime -c "import humanize" 2> import.log && tuna import.log
```

Here's `main`:

image

Here's this PR:

image

With the suggestions:

image

**Comment by dangillet** [comment]
Hi,

Thank you for the review. I had not considered the CLI aspect of humanize. That's a good point. Now a CLI user will pay the price of importing `importlib.resources` at some point obviously. But I can see how
```python
if TYPE_CHECKING:
    import os
    import pathlib
```
is really useful here.

Thank you for showing me tuna. I did not know this tool, and I've used snakeViz in the past, so it's great to see a better alternative. Also this is the first time I see `python -X importtime -c "..."` in use. I learnt something new today! :clap:

**Comment by hugovk** [comment]
I can also imagine some CLI users don't use i18n so might not run touch this. Even for code that might run more often, lazy imports can make `--help` a bit faster too.

Yeah, tuna is really nice for visualising import times along with `-X importtime` :)

Thanks for the PR!

**Pull request merged**